src/gui/properties/piecesbar.cpp

```diff
@@ -0,0 +1,325 @@
+{
+    return m_bgColor;
+}
+
+QColor PiecesBar::borderColor() const
+{
+    return m_borderColor;
+}
+
+QColor PiecesBar::pieceColor() const
+{
+    return m_pieceColor;
+}
+
+const QVector<QRgb> & PiecesBar::pieceColors() const
+{
+    return m_pieceColors;
+}
+
+QRgb PiecesBar::mixTwoColors(QRgb rgb1, QRgb rgb2, float ratio)
+{
+    int r1 = qRed(rgb1);
+    int g1 = qGreen(rgb1);
+    int b1 = qBlue(rgb1);
+
+    int r2 = qRed(rgb2);
+    int g2 = qGreen(rgb2);
+    int b2 = qBlue(rgb2);
+
+    float ratio_n = 1.0 - ratio;
+    int r = (r1 * ratio_n) + (r2 * ratio);
+    int g = (g1 * ratio_n) + (g2 * ratio);
+    int b = (b1 * ratio_n) + (b2 * ratio);
+
+    return qRgb(r, g, b);
+}
+
+namespace
+{
+#define TYPESET_FILELIST_AS_TABLE
+
+    class DetailedTooltipRenderer
+    {
+    public:
+        DetailedTooltipRenderer(QTextStream& stream, const QString &header)
+            : m_stream(stream)
+        {
+            m_stream << header
+#ifdef TYPESET_FILELIST_AS_TABLE
+                << R"(<table style="width:100%; padding: 3px; vertical-align: middle;">)";
+#else
+                << "<ul>";
+#endif
+        }
+
+        ~DetailedTooltipRenderer()
+        {
+#ifdef TYPESET_FILELIST_AS_TABLE
+            m_stream << "</table>";
+#else
+            m_stream << "</ul>";
+#endif
+        }
+
+        void operator()(const QString &size, const QString &path)
+        {
+#ifdef TYPESET_FILELIST_AS_TABLE
+            m_stream << R"(<tr><td style="white-space:nowrap">)" << size << "</td><td>" << path << "</td></tr>";
+#else
+            m_stream << "<li>[" << size << "] " << path << "</li>";
+#endif
+        }
+
+    private:
+        QTextStream &m_stream;
+    };
+}
+
+void PiecesBar::showToolTip(const QHelpEvent *e)
+{
+    if (!m_torrent)
+        return;
+
+    QString toolTipText;
+    QTextStream stream(&toolTipText, QIODevice::WriteOnly);
+    bool showDetailedInformation = QApplication::keyboardModifiers().testFlag(Qt::ShiftModifier);
+    if (!showDetailedInformation) {
+        stream << simpleToolTipText();
+        stream << '\n' << tr("Hold Shift key for detailed information");
+    }
+    else {
+        stream << "<html><body>";
+        const int imagePos = e->pos().x() - borderWidth;
+        if ((imagePos >=0) && (imagePos < m_image.width())) {
+            PieceIndexToImagePos transform {m_torrent, m_image};
+            int pieceIndex = transform.pieceIndex(e->pos().x() - borderWidth);
+            QList<int> files {m_torrent->info().fileIndicesForPiece(pieceIndex)};
+
+            DetailedTooltipRenderer renderer(stream,
+                                             files.count() > 1 ? tr("Files in this piece:") : tr("Files in these pieces:"));
+
+            const bool isFileNameCorrectionNeeded = this->isFileNameCorrectionNeeded();
+            for (int f: files) {
+                QString filePath {m_torrent->info().filePath(f)};
+                if (!isFileNameCorrectionNeeded)
```

Comment:
No we don't. But the name of the function indicates that you don't need the exclamation mark. So you need to changed the returned values or the function name to avoid confusions.